File: LeetCode/Explore/TopInterview/Amazon/Tree/101_symmetric_tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
  
private:
  bool isSameTree(TreeNode *p, TreeNode *q) {
    if (p == q) {
      return true;
    }
    if (!p || !q) {
      return false;
    }
    return (p->val == q->val) && isSameTree(p->left, q->right) && isSameTree(p->right, q->left);
  }
  
public:
    bool isSymmetric(TreeNode* root) {
      if (!root) {
        return true;
      }
      return isSameTree(root->left, root->right);
    }
};

File: LeetCode/Explore/TopInterview/Amazon/Tree/101_symmetric_tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
  
    def isSameTree(self, p: Optional[TreeNode], q: Optional[TreeNode]):
      if p == q:
        return True
      if p == None or q == None:
        return False
      return p.val == q.val and self.isSameTree(p.left, q.right) and self.isSameTree(p.right, q.left)
  
    def isSymmetric(self, root: Optional[TreeNode]) -> bool:
      if root == None:
        return True
      return self.isSameTree(root.left, root.right)
